File: sdk/python/stele/envelope.py
"""Envelope canonical encoding — mirrors pkg/attest/attest.go: Envelope.Canonical().

This file is the cryptographic contract between Python producers and the
Go operator. If the byte layout here diverges from the Go side by even
one bit, every signature this SDK produces will fail to verify.

The wire format:

    u32 len(ProducerID)        || ProducerID
    i64 TimeNanos              (big-endian)
    u32 len(Source)            || Source
    u32 len(Data)              || Data
    u32 len(PublicKey)         || PublicKey
    u32 len(AttestationType)   || AttestationType
    u32 len(EvidenceHash)      || EvidenceHash
    u32 len(QuantumPublicKey)  || QuantumPublicKey   (zero-length in classical mode)

All u32 lengths are big-endian. Strings are encoded as UTF-8 bytes
WITHOUT a null terminator. The trailing QuantumPublicKey field is
always written (zero-length when not in hybrid mode) so that adding
a quantum key in the future cannot collide with a classical signature.
"""

from __future__ import annotations

import hashlib
import json
import struct
import time
from base64 import b64encode
from dataclasses import dataclass, field
from typing import Optional


# Type discriminator for the AttestationType field. Mirrors
# pkg/attest/attest.go AttestationType constants. We support
# "software" — the platform-agnostic producer kind — and pass
# through any other string for future compatibility.
TYPE_SOFTWARE = "software"


@dataclass
class Envelope:
    """A producer-signed envelope ready to POST to /api/v0/append.

    Construct with the result of :func:`new_envelope`, then call
    :meth:`sign` with the producer's PrivateKey. The resulting envelope
    is JSON-serialisable via :meth:`to_dict`.
    """

    producer_id: str
    time_nanos: int
    source: str
    data: bytes
    public_key: bytes  # 32-byte Ed25519 pubkey
    attestation_type: str = TYPE_SOFTWARE
    evidence_hash: bytes = b""
    evidence: bytes = b""
    signature: bytes = b""
    # Hybrid (post-quantum) fields kept for forward compat. Classical
    # producers always leave these empty; the canonical bytes still
    # encode a zero-length quantum pubkey so the hybrid-downgrade
    # defence on the Go side never opens.
    quantum_public_key: bytes = b""
    quantum_signature: bytes = b""

    def canonical(self) -> bytes:
        """Return the deterministic byte sequence this envelope's signature covers."""
        return canonical_bytes(
            producer_id=self.producer_id,
            time_nanos=self.time_nanos,
            source=self.source,
            data=self.data,
            public_key=self.public_key,
            attestation_type=self.attestation_type,
            evidence_hash=self.evidence_hash,
            quantum_public_key=self.quantum_public_key,
        )

    def hash(self) -> bytes:
        """SHA-256 of the canonical bytes — used by the operator's replay-protection table."""
        return hashlib.sha256(self.canonical()).digest()

    def sign(self, priv) -> None:
        """Sign this envelope in place with ``priv`` (a :class:`PrivateKey`)."""
        if not self.public_key:
            self.public_key = priv.public_bytes()
        self.signature = priv.sign(self.canonical())

    def to_dict(self) -> dict:
        """Serialise to a JSON-friendly dict matching the Go side's struct tags."""
        out = {
            "producer_id": self.producer_id,
            "time_ns": self.time_nanos,
            "source": self.source,
            "data": _b64(self.data),
            "public_key": _b64(self.public_key),
            "attestation_type": self.attestation_type,
            "signature": _b64(self.signature),
        }
        if self.evidence_hash:
            out["evidence_hash"] = _b64(self.evidence_hash)
        if self.evidence:
            out["evidence"] = _b64(self.evidence)
        if self.quantum_public_key:
            out["quantum_public_key"] = _b64(self.quantum_public_key)
        if self.quantum_signature:
            out["quantum_signature"] = _b64(self.quantum_signature)
        return out


def canonical_bytes(
    *,
    producer_id: str,
    time_nanos: int,
    source: str,
    data: bytes,
    public_key: bytes,
    attestation_type: str = TYPE_SOFTWARE,
    evidence_hash: bytes = b"",
    quantum_public_key: bytes = b"",
) -> bytes:
    """Pure function: produce the canonical bytes for the given envelope fields.

    Exposed so callers can independently confirm their canonicalisation
    matches the Go side. Use :meth:`Envelope.canonical` in normal flow.
    """
    out = bytearray()
    _put_bytes(out, producer_id.encode("utf-8"))
    out += struct.pack(">q", time_nanos)
    _put_bytes(out, source.encode("utf-8"))
    _put_bytes(out, data)
    _put_bytes(out, public_key)
    _put_bytes(out, attestation_type.encode("utf-8"))
    _put_bytes(out, evidence_hash)
    _put_bytes(out, quantum_public_key)
    return bytes(out)


def new_envelope(
    producer_id: str,
    public_key: bytes,
    source: str,
    data: bytes,
    *,
    attestation_type: str = TYPE_SOFTWARE,
    time_nanos: Optional[int] = None,
) -> Envelope:
    """Construct an unsigned envelope ready to be passed to :meth:`Envelope.sign`."""
    if time_nanos is None:
        time_nanos = time.time_ns()
    return Envelope(
        producer_id=producer_id,
        time_nanos=time_nanos,
        source=source,
        data=bytes(data),
        public_key=bytes(public_key),
        attestation_type=attestation_type,
    )


# ---- helpers ----


def _put_bytes(buf: bytearray, b: bytes) -> None:
    """Length-prefixed (u32 big-endian) byte string, matching Go's binary.BigEndian.PutUint32 + append."""
    buf += struct.pack(">I", len(b))
    buf += b


def _b64(b: bytes) -> str:
    return b64encode(b).decode("ascii")


# Used by the producer module to round-trip envelopes through JSON
# (e.g. when persisting unsent entries to a queue).
def envelope_to_json(env: Envelope) -> str:
    return json.dumps(env.to_dict(), separators=(",", ":"), sort_keys=True)

File: sdk/typescript/src/envelope.ts
// Envelope canonical encoding — mirrors pkg/attest/attest.go: Envelope.Canonical().
//
// The CRITICAL property: for the same inputs, the bytes here are
// byte-identical to what the Go side produces. If you change the
// layout, every signature this SDK produces will fail to verify.
//
// Wire format:
//   u32 len(ProducerID)        || ProducerID
//   i64 TimeNanos              (big-endian, signed)
//   u32 len(Source)            || Source
//   u32 len(Data)              || Data
//   u32 len(PublicKey)         || PublicKey
//   u32 len(AttestationType)   || AttestationType
//   u32 len(EvidenceHash)      || EvidenceHash
//   u32 len(QuantumPublicKey)  || QuantumPublicKey   (zero-length classical mode)

export const TYPE_SOFTWARE = "software";

/** A producer-signed envelope ready to POST to /api/v0/append. */
export interface Envelope {
  producerId: string;
  timeNanos: bigint;
  source: string;
  data: Uint8Array;
  publicKey: Uint8Array;
  attestationType: string;
  evidenceHash?: Uint8Array;
  evidence?: Uint8Array;
  signature?: Uint8Array;
  quantumPublicKey?: Uint8Array;
  quantumSignature?: Uint8Array;
}

/** Inputs to canonicalBytes — exactly the fields covered by the signature. */
export interface CanonicalInputs {
  producerId: string;
  timeNanos: bigint;
  source: string;
  data: Uint8Array;
  publicKey: Uint8Array;
  attestationType?: string;
  evidenceHash?: Uint8Array;
  quantumPublicKey?: Uint8Array;
}

/** Pure function: produce the canonical bytes for the given envelope fields. */
export function canonicalBytes(inp: CanonicalInputs): Uint8Array {
  const enc = new TextEncoder();
  const parts: Uint8Array[] = [];

  parts.push(putBytes(enc.encode(inp.producerId)));
  parts.push(putInt64BE(inp.timeNanos));
  parts.push(putBytes(enc.encode(inp.source)));
  parts.push(putBytes(inp.data));
  parts.push(putBytes(inp.publicKey));
  parts.push(putBytes(enc.encode(inp.attestationType ?? TYPE_SOFTWARE)));
  parts.push(putBytes(inp.evidenceHash ?? new Uint8Array()));
  parts.push(putBytes(inp.quantumPublicKey ?? new Uint8Array()));

  return concat(parts);
}

/** SHA-256 of the canonical bytes — used by the operator's replay-protection table. */
export async function envelopeHash(env: Envelope): Promise<Uint8Array> {
  const canonical = canonicalBytes({
    producerId: env.producerId,
    timeNanos: env.timeNanos,
    source: env.source,
    data: env.data,
    publicKey: env.publicKey,
    attestationType: env.attestationType,
    evidenceHash: env.evidenceHash,
    quantumPublicKey: env.quantumPublicKey,
  });
  const buf = await getCrypto().subtle.digest(
    "SHA-256",
    canonical as unknown as BufferSource,
  );
  return new Uint8Array(buf);
}

/** Serialise an envelope to a JSON-friendly object matching the Go struct tags. */
export function envelopeToWire(env: Envelope): Record<string, unknown> {
  if (!env.signature || env.signature.length === 0) {
    throw new Error("envelopeToWire: signature missing — call signEnvelope first");
  }
  const out: Record<string, unknown> = {
    producer_id: env.producerId,
    time_ns: Number(env.timeNanos),
    source: env.source,
    data: b64(env.data),
    public_key: b64(env.publicKey),
    attestation_type: env.attestationType,
    signature: b64(env.signature),
  };
  if (env.evidenceHash && env.evidenceHash.length > 0) {
    out.evidence_hash = b64(env.evidenceHash);
  }
  if (env.evidence && env.evidence.length > 0) {
    out.evidence = b64(env.evidence);
  }
  if (env.quantumPublicKey && env.quantumPublicKey.length > 0) {
    out.quantum_public_key = b64(env.quantumPublicKey);
  }
  if (env.quantumSignature && env.quantumSignature.length > 0) {
    out.quantum_signature = b64(env.quantumSignature);
  }
  return out;
}

// ---- helpers ----

function putBytes(b: Uint8Array): Uint8Array {
  const len = new ArrayBuffer(4);
  new DataView(len).setUint32(0, b.length, false); // big-endian
  const out = new Uint8Array(4 + b.length);
  out.set(new Uint8Array(len), 0);
  out.set(b, 4);
  return out;
}

function putInt64BE(n: bigint): Uint8Array {
  const buf = new ArrayBuffer(8);
  new DataView(buf).setBigInt64(0, n, false); // big-endian
  return new Uint8Array(buf);
}

function concat(parts: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const p of parts) total += p.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const p of parts) {
    out.set(p, off);
    off += p.length;
  }
  return out;
}

/** Base64-encode with no line wrapping, no padding stripping. */
export function b64(b: Uint8Array): string {
  // Node and modern browsers both expose btoa/atob OR Buffer; we
  // pick the route that exists.
  if (typeof Buffer !== "undefined") {
    return Buffer.from(b).toString("base64");
  }
  let s = "";
  for (let i = 0; i < b.length; i++) {
    s += String.fromCharCode(b[i]!);
  }
  return btoa(s);
}

export function b64decode(s: string): Uint8Array {
  if (typeof Buffer !== "undefined") {
    return new Uint8Array(Buffer.from(s, "base64"));
  }
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) {
    out[i] = bin.charCodeAt(i);
  }
  return out;
}

/** Get the WebCrypto SubtleCrypto in both Node ≥18 and browsers. */
export function getCrypto(): Crypto {
  if (typeof globalThis !== "undefined" && globalThis.crypto?.subtle) {
    return globalThis.crypto;
  }
  // Older Node fallback path.
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  return require("node:crypto").webcrypto as Crypto;
}
